### ISSUE TITLE
"no-use-before-define" should have an option to ignore type declarations

### ISSUE THREAD

**Issue by pasha-mf** [eslint/eslint#4867]
1. `v1.10.3`
2.  Using parser "babel-eslint", rule `"no-use-before-define": 1` and the following test file:
```
type FooBar = Foo | Bar;
type Foo = { foo: string };
type Bar = { bar: number };
```
produces errors that Foo & Bar are used before they are defined. However type declarations are considered to be hoisted to the top of the file, so this syntax is valid.
3.
```
   1:15  warning  "Foo" was used before it was defined                       no-use-before-define
   1:21  warning  "Bar" was used before it was defined                       no-use-before-define
```
4. Expected no errors to be reported.

**Comment by eslintbot** [comment]
@pasha-mf Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Please file a bug with babel-eslint.
